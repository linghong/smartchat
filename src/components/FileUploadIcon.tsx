import React, { useRef, ChangeEvent, FC } from 'react';
import { Paperclip } from 'lucide-react';

interface FileUploadIconProps {
  onFileUpload: (file: File) => void;
  isDisabled: boolean;
  accept?: string; // Optional prop to allow for customizable accepted file types
}

const FileUploadIcon: FC<FileUploadIconProps> = ({
  onFileUpload,
  isDisabled,
  accept = 'image/*, .pdf, .html, .xml, .docx, .txt, .yml, .yaml, .tsx, .ts, .js, .cjs, .coffee, .css, .scss, .sass, less, .xsl, .xlsx,  .jsx, .java, .py, .php, .rb, .erb, .rake, .slim, .cpp, .c, .go, .rs, .log, .cjs, .snap, .sh, .md, .csv, .html, .json, .toml, .sh,  .map,  .xml, .lock, .mod, .metal, .sum, .h, .m, .asp, .net, .dll, .cs, .vb, .fs, .config, .resx, .csproj, .sln'
  // not supported:.rtf, .psd, .eps, .pyc, .ppt, .pptx, .icns, .svg, .cu, .icons, .bin, .exe, .env.example, etc.
}) => {
  const fileInputRef = useRef<HTMLInputElement>(null);

  // Trigger the file input when onKeyDown is fired
  const handleIconClick = () => {
    fileInputRef.current?.click();
  };

  const handleFileChange = (event: ChangeEvent<HTMLInputElement>) => {
    const file = event.target.files ? event.target.files[0] : null;
    if (!file) return;

    onFileUpload(file);
  };

  return (
    <label
      className="flex items-center justify-center  text-white font-bold px-2 rounded cursor-pointer"
      tabIndex={0}
      onKeyDown={e => e.key === 'Enter' && handleIconClick()}
    >
      <input
        type={`${isDisabled ? 'none' : 'file'}`}
        ref={fileInputRef}
        onChange={handleFileChange}
        style={{ display: 'none' }}
        accept={accept}
        aria-label="Upload file"
      />
      <Paperclip
        size={26}
        style={{
          background: '#fafafb',
          cursor: 'pointer',
          color: '#2c2e2f',
          border: 'solid #d7dadf 1px'
        }}
        aria-hidden="true" // Hides the icon from screen readers
      />
    </label>
  );
};

export default FileUploadIcon;
