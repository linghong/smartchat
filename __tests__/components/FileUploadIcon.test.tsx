import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import FileUploadIcon from '@/src/components/FileUploadIcon';

describe('FileUploadIcon', () => {
  const onFileUpload = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the upload icon', () => {
    render(<FileUploadIcon onFileUpload={onFileUpload} isDisabled={false} />);

    const icon = screen.getByLabelText('Upload file');
    expect(icon).toBeInTheDocument();
  });

  it('should trigger file input click when icon is clicked', () => {
    render(<FileUploadIcon onFileUpload={onFileUpload} isDisabled={false} />);

    const icon = screen.getByLabelText('Upload file');
    fireEvent.click(icon);

    const fileInput = screen.getByLabelText('Upload file');
    expect(fileInput).toBeInTheDocument();
    expect(fileInput).toHaveProperty('type', 'file');
  });

  it('should trigger file input click when Enter key is pressed', () => {
    render(<FileUploadIcon onFileUpload={onFileUpload} isDisabled={false} />);

    const icon = screen.getByLabelText('Upload file');
    fireEvent.keyDown(icon, { key: 'Enter', code: 'Enter' });

    const fileInput = screen.getByLabelText('Upload file');
    expect(fileInput).toBeInTheDocument();
    expect(fileInput).toHaveProperty('type', 'file');
  });

  it('should call onFileUpload when a file is selected and not disabled', () => {
    render(<FileUploadIcon onFileUpload={onFileUpload} isDisabled={false} />);

    const fileInput = screen.getByLabelText('Upload file') as HTMLInputElement;
    const file = new File(['dummy content'], 'example.txt', {
      type: 'text/plain'
    });
    fireEvent.change(fileInput, { target: { files: [file] } });

    expect(onFileUpload).toHaveBeenCalledTimes(1);
    expect(onFileUpload).toHaveBeenCalledWith(file);
  });

  it('should not allow file upload when disabled', () => {
    render(<FileUploadIcon onFileUpload={onFileUpload} isDisabled={true} />);

    const fileInput = screen.getByLabelText('Upload file') as HTMLInputElement;
    expect(fileInput).toHaveAttribute('type', 'none');

    const file = new File(['dummy content'], 'example.txt', {
      type: 'text/plain'
    });
    fireEvent.change(fileInput, { target: { files: [file] } });

    expect(onFileUpload).not.toHaveBeenCalled();
  });

  it('should accept various file types', () => {
    render(<FileUploadIcon onFileUpload={onFileUpload} isDisabled={false} />);

    const fileInput = screen.getByLabelText('Upload file') as HTMLInputElement;
    expect(fileInput).toHaveAttribute(
      'accept',
      expect.stringContaining('image/*, .pdf, .html, .xml, .docx, .txt')
    );
  });

  it('should allow custom accept attribute', () => {
    const customAccept = '.csv, .xlsx';
    render(
      <FileUploadIcon
        onFileUpload={onFileUpload}
        isDisabled={false}
        accept={customAccept}
      />
    );

    const fileInput = screen.getByLabelText('Upload file') as HTMLInputElement;
    expect(fileInput).toHaveAttribute('accept', customAccept);
  });

  it('should match snapshot when enabled', () => {
    const { asFragment } = render(
      <FileUploadIcon onFileUpload={onFileUpload} isDisabled={false} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when disabled', () => {
    const { asFragment } = render(
      <FileUploadIcon onFileUpload={onFileUpload} isDisabled={true} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
