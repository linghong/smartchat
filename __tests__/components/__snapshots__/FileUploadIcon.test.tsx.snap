// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploadIcon should match snapshot when disabled 1`] = `
<DocumentFragment>
  <label
    class="flex items-center justify-center  text-white font-bold px-2 rounded cursor-pointer"
    tabindex="0"
  >
    <input
      accept="image/*, .pdf, .html, .xml, .docx, .txt, .yml, .yaml, .tsx, .ts, .js, .cjs, .coffee, .css, .scss, .sass, less, .xsl, .xlsx,  .jsx, .java, .py, .php, .rb, .erb, .rake, .slim, .cpp, .c, .go, .rs, .log, .cjs, .snap, .sh, .md, .csv, .html, .json, .toml, .sh,  .map,  .xml, .lock, .mod, .metal, .sum, .h, .m, .asp, .net, .dll, .cs, .vb, .fs, .config, .resx, .csproj, .sln"
      aria-label="Upload file"
      style="display: none;"
      type="none"
    />
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      style="cursor: pointer; background: rgb(123, 123, 220); border: 1px solid black;"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 15V18H24V20H21V23H19V20H16V18H19V15H21ZM21.0082 3C21.556 3 22 3.44495 22 3.9934L22.0007 13.3417C21.3749 13.1204 20.7015 13 20 13V5H4L4.001 19L13.2929 9.70715C13.6528 9.34604 14.22 9.31823 14.6123 9.62322L14.7065 9.70772L18.2521 13.2586C15.791 14.0069 14 16.2943 14 19C14 19.7015 14.1204 20.3749 14.3417 21.0007L2.9918 21C2.44405 21 2 20.5551 2 20.0066V3.9934C2 3.44476 2.45531 3 2.9918 3H21.0082ZM8 7C9.10457 7 10 7.89543 10 9C10 10.1046 9.10457 11 8 11C6.89543 11 6 10.1046 6 9C6 7.89543 6.89543 7 8 7Z"
      />
    </svg>
  </label>
</DocumentFragment>
`;

exports[`FileUploadIcon should match snapshot when enabled 1`] = `
<DocumentFragment>
  <label
    class="flex items-center justify-center  text-white font-bold px-2 rounded cursor-pointer"
    tabindex="0"
  >
    <input
      accept="image/*, .pdf, .html, .xml, .docx, .txt, .yml, .yaml, .tsx, .ts, .js, .cjs, .coffee, .css, .scss, .sass, less, .xsl, .xlsx,  .jsx, .java, .py, .php, .rb, .erb, .rake, .slim, .cpp, .c, .go, .rs, .log, .cjs, .snap, .sh, .md, .csv, .html, .json, .toml, .sh,  .map,  .xml, .lock, .mod, .metal, .sum, .h, .m, .asp, .net, .dll, .cs, .vb, .fs, .config, .resx, .csproj, .sln"
      aria-label="Upload file"
      style="display: none;"
      type="file"
    />
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="24"
      stroke="currentColor"
      stroke-width="0"
      style="cursor: pointer; background: rgb(123, 123, 220); border: 1px solid black;"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 15V18H24V20H21V23H19V20H16V18H19V15H21ZM21.0082 3C21.556 3 22 3.44495 22 3.9934L22.0007 13.3417C21.3749 13.1204 20.7015 13 20 13V5H4L4.001 19L13.2929 9.70715C13.6528 9.34604 14.22 9.31823 14.6123 9.62322L14.7065 9.70772L18.2521 13.2586C15.791 14.0069 14 16.2943 14 19C14 19.7015 14.1204 20.3749 14.3417 21.0007L2.9918 21C2.44405 21 2 20.5551 2 20.0066V3.9934C2 3.44476 2.45531 3 2.9918 3H21.0082ZM8 7C9.10457 7 10 7.89543 10 9C10 10.1046 9.10457 11 8 11C6.89543 11 6 10.1046 6 9C6 7.89543 6.89543 7 8 7Z"
      />
    </svg>
  </label>
</DocumentFragment>
`;
